core/src/com/google/inject/internal/BytecodeGen.java

```diff
@@ -16,412 +16,147 @@
-    net.sf.cglib.proxy.Enhancer enhancer = new net.sf.cglib.proxy.Enhancer();
-    enhancer.setSuperclass(type);
-    enhancer.setUseFactory(false);
-    if (visibility == Visibility.PUBLIC) {
-      enhancer.setClassLoader(getClassLoader(type));
-    }
-    enhancer.setNamingPolicy(ENHANCER_NAMING_POLICY);
-    logger.fine("Loading " + type + " Enhancer with " + enhancer.getClassLoader());
-    return enhancer;
-  }
-  /*end[AOP]*/
-
   /**
-   * The required visibility of a user's class from a Guice-generated class. Visibility of
-   * package-private members depends on the loading classloader: only if two classes were loaded by
-   * the same classloader can they see each other's package-private members. We need to be careful
-   * when choosing which classloader to use for generated classes. We prefer our bridge classloader,
-   * since it's OSGi-safe and doesn't leak permgen space. But often we cannot due to visibility.
+   * Prepares the class containing the given member for fast invocation using bytecode generation.
    */
-  public enum Visibility {
-
-    /**
-     * Indicates that Guice-generated classes only need to call and override public members of the
-     * target class. These generated classes may be loaded by our bridge classloader.
-     */
-    PUBLIC {
-      @Override
-      public Visibility and(Visibility that) {
-        return that;
-      }
-    },
-
-    /**
-     * Indicates that Guice-generated classes need to call or override package-private members.
-     * These generated classes must be loaded in the same classloader as the target class. They
-     * won't work with OSGi, and won't get garbage collected until the target class' classloader is
-     * garbage collected.
-     */
-    SAME_PACKAGE {
-      @Override
-      public Visibility and(Visibility that) {
-        return this;
-      }
-    };
-
-    public static Visibility forMember(Member member) {
-      if ((member.getModifiers() & (Modifier.PROTECTED | Modifier.PUBLIC)) == 0) {
-        return SAME_PACKAGE;
-      }
-
-      Class[] parameterTypes;
-      if (member instanceof Constructor) {
-        parameterTypes = ((Constructor) member).getParameterTypes();
-      } else {
-        Method method = (Method) member;
-        if (forType(method.getReturnType()) == SAME_PACKAGE) {
-          return SAME_PACKAGE;
-        }
-        parameterTypes = method.getParameterTypes();
-      }
-
-      for (Class<?> type : parameterTypes) {
-        if (forType(type) == SAME_PACKAGE) {
-          return SAME_PACKAGE;
-        }
-      }
-
-      return PUBLIC;
-    }
-
-    public static Visibility forType(Class<?> type) {
-      return (type.getModifiers() & (Modifier.PROTECTED | Modifier.PUBLIC)) != 0
-          ? PUBLIC
-          : SAME_PACKAGE;
+  private static Function<String, BiFunction> fastClass(Executable member) {
+    Class<?> hostClass = member.getDeclaringClass();
+    if (hostClass.getSimpleName().contains(ENHANCER_BY_GUICE_MARKER)) {
```

Comment:
This can happen when the enhanced class uses method-injection. In such cases the fast-class generated for the original host class will also work for the enhanced class, so this avoids generating an unnecessary fast-class.

Reply:
Consider adding a comment here to mention that.

Reply:
Did more testing and this can safely be removed: intercepted setters already use the fast-class of the original declaring class, so nothing was actually triggering this: d0f6176